nvtabular/io.py

```diff
@@ -456,74 +472,111 @@ def _write_thread(self):
                 if item is self._eod:
                     break
                 idx, data = item
-                ones = np.array(([1] * data.shape[0]), dtype=np.intc)
                 with self.write_locks[idx]:
-                    df = data[self.column_names].to_pandas().astype(np.single)
-                    for i in range(len(self.cats)):
-                        df["___" + str(i) + "___" + self.cats[i]] = ones
-                        df[self.cats[i]] = data[self.cats[i]].to_pandas().astype(np.longlong)
-
-                    self.writers[idx].write(df.to_numpy().tobytes())
+                    self.data_writers[idx].write_table(data)
             finally:
                 self.queue.task_done()
 
     @annotate("add_data", color="orange", domain="nvt_python")
     def add_data(self, gdf):
-        # get slice info
-        int_slice_size = gdf.shape[0] // self.num_out_files
-        slice_size = int_slice_size if gdf.shape[0] % int_slice_size == 0 else int_slice_size + 1
+        arr = cp.arange(len(gdf))
+        cp.random.shuffle(arr)
+        np.random.shuffle(self.b_idxs)
 
-        for x in range(self.num_out_files):
-            start = x * slice_size
-            end = start + slice_size
-            # check if end is over length
-            end = end if end <= gdf.shape[0] else gdf.shape[0]
-            to_write = gdf.iloc[start:end]
-            self.num_samples[x] = self.num_samples[x] + to_write.shape[0]
-            self.queue.put((x, to_write))
+        Writer.add_data(self, gdf, arr)
```

Comment:
I don't think this will work - the super class won't be able to write out the file since it's missing some implementation details.

I was thinking that we would pass another writer to the Shuffler object in the ```__init__``` call, and then call that writers add_data method after shuffling.

I think we should also move the threadpool code out of the base class, since otherwise we'd have multiple threadpools in this case.


The class hierarchy in this case would look like:

```
Writer: defines the interface required (add_data/close methods)

    -> Shuffler : shuffles a dataframe and passes the shuffled DF to another writer

    -> ThreadedWriter: adds a threadpool for parallelizing writers
          -> ParquetWriter: writes out a parquet dataset
          -> HugeCTRWriter: writes out a dataset in the hugectr binary format
```